### PR TITLE
search: fix type:repo behavior

### PR DESCRIPTION
**Pull request by rvantonder**
Addresses https://github.com/sourcegraph/sourcegraph/issues/19239. 

This is a fairly complex and intentional behavioral change to `type:repo`.  Considering it was basically broken for 3 months on Sourcegraph.com and hindering unfamiliar users I'm pretty aggressive about the behavioral changes here.

## What did `type:repo` do before?

When it did work, it would behave roughly as:

- `type:repo foo bar` means `repo:"foo bar"`. I.e., if a user typed `type:repo facebook react` they are basically guaranteed to not get any results on Sourcegraph.com, since hardly any repos on GH contain a space (zero?)

- `type:repo foo bar` in regex mode means `repo:foo.*bar`, as we take the convention of interpolating `.*` for spaces in regex mode.

## What does `type:repo` do now?

There's a comment in the code, but basically:

- `type:repo any thing` is translated to `repo:any repo:thing`. This means we will do an `and` search for space-separated patterns (as opposed to content searches, where spaces are significant for ordering). The `and` search is more suited when the user expects to search for repositories containing all these keywords. `type:repo facebook react` will give a lot of results (not necessarily the order we want, but we'll get hits).

- `type:repo a or not b` is translated to `repo:a or -repo:b`. Basically, any `and` or `or` operator is preserved when translating patterns with a `repo:` prefix, and additionally, `not` works as is

- caveat: because of the complex behavior must restrict the use of `type:repo` to only allow `type:repo` in isolation, it cannot be combined with other `type:` queries. I.e., `(type:repo or type:file) an example` is not supported. This because we want `type:repo` to interpret "the rest" of the query at translation stage different to any other query shape, where other query shapes are understood as ordered patterns as-is, or otherwise rely on whatever inlined behavior we have elsewhere in the backend). It's not that it's not supportable. It's that it would take me 2 weeks+ to modify parsing so that it now takes into account context from something like `type` within a particular scope, which I would prefer not exist in the first place.

### CODE REVIEW

File: internal/search/query/validate_test.go [hunk 1]
```diff
@@ -346,55 +350,3 @@ func TestContainsRefGlobs(t *testing.T) {
 		})
 	}
 }
-
-func TestHasTypeRepo(t *testing.T) {
```
Comment (rvantonder) [inline internal/search/query/validate_test.go:350]: This test is no longer useful

File: internal/search/query_converter_test.go [hunk 2]
```diff
@@ -46,15 +46,15 @@ func TestToTextPatternInfo(t *testing.T) {
...
 
-	autogold.Want("05", `{"Pattern":"sgtest/mux","IsNegated":false,"IsRegExp":true,"IsStructuralPat":false,"CombyRule":"","IsWordMatch":false,"IsCaseSensitive":false,"FileMatchLimit":30,"Index":"yes","Select":{"Type":"","Fields":null},"IncludePatterns":null,"ExcludePattern":"","FilePatternsReposMustInclude":null,"FilePatternsReposMustExclude":null,"PathPatternsAreCaseSensitive":false,"PatternMatchesContent":false,"PatternMatchesPath":false,"Languages":null}`).Equal(t, test(`type:repo sgtest/mux fork:yes`))
+	autogold.Want("05", `{"Pattern":"","IsNegated":false,"IsRegExp":true,"IsStructuralPat":false,"CombyRule":"","IsWordMatch":false,"IsCaseSensitive":false,"FileMatchLimit":30,"Index":"yes","Select":{"Type":"","Fields":null},"IncludePatterns":null,"ExcludePattern":"","FilePatternsReposMustInclude":null,"FilePatternsReposMustExclude":null,"PathPatternsAreCaseSensitive":false,"PatternMatchesContent":false,"PatternMatchesPath":false,"Languages":null}`).Equal(t, test(`type:repo sgtest/mux fork:yes`))
```
Comment (rvantonder) [inline internal/search/query_converter_test.go:57]: `Pattern` gets nuked as expected because these are `type:repo` queries.

**Comment by limitedmage** [comment]
> * caveat: because of the complex behavior must restrict the use of `type:repo` to only allow `type:repo` in isolation, it cannot be combined with other `type:` queries. I.e., `(type:repo or type:file) an example` is not supported. This because we want `type:repo` to interpret "the rest" of the query at translation stage different to any other query shape, where other query shapes are understood as ordered patterns as-is, or otherwise rely on whatever inlined behavior we have elsewhere in the backend). It's not that it's not supportable. It's that it would take me 2 weeks+ to modify parsing so that it now takes into account context from something like `type` within a particular scope, which I would prefer not exist in the first place.

What happens if the user tries to do a search like this? Will the query still work with the old behavior, or will there be an error? Will this break the default search which is essentially `type:repo or type:file or type:path` if you omit `type:`? How will this affect the plan to add the checkbox UI for type in the sidebar?

**Comment by rvantonder** [comment]
> What happens if the user tries to do a search like this? Will the query still work with the old behavior, or will there be an error?

You get this:

```
`type:repo` may not be used with any other `type:` parameter. The query contains both `type:repo` and `type:%s`
```

> Will this break the default search which is essentially type:repo or type:file or type:path if you omit type:? 

It doesn't break any default search results type, which we define in the backend. Absence of `type:` has no change in behavior. It does mean that a user cannot explicitly compose `type` using operators like `type:repo or type:file or type:path` if it includes `type:repo`. Note that in the default search the repos search (`type:repo`) is, essentially, useless if the pattern contains a space, as per above.


> How will this affect the plan to add the checkbox UI for type in the sidebar?

We are kinda ❚❚❚❚❚❚❚ in that state even before this change based on what `type:repo` does. Let's go again with a user entering a query like `facebook react`, and they select `type:repo`. Again this won't realistically match any repositories anyway, ever. So, adding `type:repo` to any other `type` is a useless action as soon as the pattern contains a space, just like it is useless when we search all types by default. We would introduce the checkbox to make it possible to select, but the functionality would be fake for any pattern containing space. You will only see results for other types. 

With this change, _at least_ a `type:repo` is in a state where it returns results and can break from the strict structure of the pattern. Granted, this change closes the ability to explicitly select `type:repo` along with a different `type`. For now. I can see how it would be easily possible to allow `type:repo` along with any other `type`, if the pattern is contiguous (no space). This is not a problem and I can make it work. In fact, it is fairly straightforward to make things work where we allow `type:repo type:file type:content` (as we do currently, as a global addition). It would just mean creating multiple search expressions and union-ing them together. So conceptually, we are "unblocked". I am basically in favor of having `type:XYZ` _interpret_ the pattern parts of a query _differently_ in a way that makes sense for `type:XYZ` to return results that users are likely interested in. What we cannot have any time soon, unless I get 2 weeks, is allowing `type` to scope expressions, like `(type:repo foo bar) or (type:file baz qux)`.

Because `type`  UI checkboxes wouldn't have the know-how to create scoped expressions, they would just assume "append to global query" anyway so I foresee the UI interactions to be conflict-free. At the same time, I wish `type` didn't exist as a visible query member in the webapp. That is, IMO, the crux of the problem and the complexity keeps encroaching on the query string, when the complexity should be handled by UI and layout.

**Comment by rvantonder** [comment]
Filed https://github.com/sourcegraph/sourcegraph/issues/20572 for follow on things.

**Comment by limitedmage** [comment]
Thanks for the thorough explanation! FYI @rrhyne

**Comment by keegancsmith** [comment]
This is a bit confusing to me. Our default types when doing a search are `file`, `path` and `repo`. ie if I just type `sourcegraph/sourcegraph` as my search, one of the results should be the repo `github.com/sourcegraph/sourcegraph` (and it is today). But now when I explicitly specify `type:repo` it changes how we interpret pattern? Do we still keep the old behaviour of also searching repo names with pattern in a default search?

**Comment by rvantonder** [comment]
@keegancsmith I understand some of this is a bit confusing, because it's doing two things at once (maybe too much). I'm going to split those concerns. The part about making `type:repo <pattern>` alias to `repo:pattern` is here to stay, and will come in a different PR. The part about behavioral changes I'm proposing to `type:repo` I believe make a lot of sense, but I need to give more context and a comprehensive solution. I started writing a long response here, but decided it deserves an RFC to document what I'm seeking to achieve here.

**Comment by rrhyne** [comment]
@rvantonder thanks for working on this. I just want to note that I think we should avoid focusing on the spacing issue presented by queries like `facebook react type:repo` before we try cheaper routes to educating new users about our literal search model.

**Pull request closed without merging**